sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala

```diff
@@ -1084,6 +1087,32 @@ class ParquetIOSuite extends QueryTest with ParquetTest with SharedSparkSession
       }
     }
   }
+
+  test("SPARK-37035: Improve error message when use parquet vectorized reader") {
```

Comment:
If it's too hard to come up with an end-to-end test (can't construct the corrupted parquet file), I'm fine with this UT.

Reply:
I spent sometime looking for a e2e test but couldn't figure out one 🤷 . The errors with a corrupted Parquet file would either be caught when initializing the Parquet dictionary (for instance, when trying to initialize a Parquet INT64 dictionary with INT32 type), or in Spark when checking Spark schema versus Parquet schema. 

It's be interesting to know how you encountered this error originally @AngersZhuuuu . Which Spark version you were using?

Reply:
> I spent sometime looking for a e2e test but couldn't figure out one 🤷 . The errors with a corrupted Parquet file would either be caught when initializing the Parquet dictionary (for instance, when trying to initialize a Parquet INT64 dictionary with INT32 type), or in Spark when checking Spark schema versus Parquet schema.
> 
> It's be interesting to know how you encountered this error originally @AngersZhuuuu . Which Spark version you were using?

This situation has been encountered several times, and each time I repeated the test to determine which partition's file had a problem with the data, and then rewritten the data to solve it, so I did this PR.

I don't know why this happen. Hmm last time when I want to save the file with bad data,  the data have been replaced by our user. So I can't provide you a such file now. Maybe next time ==

Reply:
@sadikovi is it possible to create a valid parquet file and flip some bits to simulate a disk/network error?

Reply:
Oh, you could do something like this: 
```
(0 until 1000).map(_ => (1, "abc")).toDF("a", "b").coalesce(1).write.option("compression", "none").mode("overwrite").parquet("file:/tmp/test.parquet")
```

```
import org.apache.hadoop.fs._

val path = new Path("file:/tmp/test.parquet")
val fs = path.getFileSystem(spark.sessionState.newHadoopConf)
val status = fs.listStatus(path).filter(_.getPath.getName.endsWith(".parquet")).head
val in = fs.open(status.getPath)
val arr = new Array[Byte](status.getLen.toInt)
in.readFully(arr)
in.close
for (i <- 27 + 52 until 27 + 52 + 2) arr(i) = 12.toByte
val out = fs.create(new Path("file:/tmp/test-bad.parquet"), true)
out.write(arr)
out.close
```

```
spark.conf.set("parquet.page.verify-checksum.enabled", "false")
spark.read.parquet("file:/tmp/test-bad.parquet")
```

Caused by: RuntimeException: java.io.IOException: could not read page Page [bytes.size=4, valueCount=1, uncompressedSize=4] in col [a] required int32 a
Caused by: IOException: could not read page Page [bytes.size=4, valueCount=1, uncompressedSize=4] in col [a] required int32 a
Caused by: IOException: FAILED_TO_UNCOMPRESS(5)

You can also disable compression to actually check the decoding issues:

Caused by: ParquetDecodingException: Failed to read from input stream
Caused by: EOFException:

Reply:
You can hardcode the bytes to flip, I think it is fine but you might need to generate large enough file for the byte flip to be meaningful.

If you would like to know the parquet page offsets, I recommend using https://github.com/sadikovi/parquet-metadata-explorer, like this:
```
display(spark.read.format("metadata").option("level", "page").load("file:/tmp/test.parquet"))
```

It will give you all of the offsets you need.